### PR TITLE
Add an overwrite argument to the Writer to add the possibility to allow for overwriting existing file.

### PR DESCRIPTION
**Pull request by mscheltienne**
Hi! I would like to propose a small additional feature: the possibility to overwrite an existing file. This would be handy as I am adding the `overwrite` argument to MNE export function in this PR: https://github.com/mne-tools/mne-python/pull/9975

For now, the PR will delete existing files if `overwrite` is set to True, which is not really a clean way of doing it. The possibility to set `overwrite` upstream would be appreciated.

You can find the minimal changes I think are necessary to add this feature. I am not familiar with `mffpy `and I might have overlooked something. Please let me know!

### CODE REVIEW

**Comment by jusjusjus** [comment]
Hi @mscheltienne, your suggestion makes sense to me.  Thanks.  However, I'm worried that as a result of your implementation some, e.g., Event files wouldn't be overwritten, but just remained astray from the old MFF.  I propose to cleanly unlink the whole MFF dir if it already exists.  What do you think?

**Comment by mscheltienne** [comment]
@jusjusjus Glad to hear that you like it. 

As I expected, I don't know enough about `mffpy` to cleanly implement it. You know better than me how to properly overwrite the writer output; so you are definitely right. 
I will check tomorrow more in detail to figure out what is created by the writer and what should be removed, and I'll update the proposal accordingly.

**Comment by mscheltienne** [comment]
What do you mean by `unlink`? I just had a quick second look at the writer code, and your remark is sound.
How about:

If `filename` already exists and overwrite is `True`: list the files in `mffdir`; overwrite the one to overwrite and delete the one that were not overwritten?

**Comment by jusjusjus** [comment]
@ephathaway, let's discuss how this API should look like.

**Comment by ephathaway** [comment]
@mscheltienne, @jusjusjus, so sorry I am just now responding to this.

> If `filename` already exists and overwrite is `True`: list the files in `mffdir`; overwrite the one to overwrite and delete the one that were not overwritten?

This sounds like a good solution. We could add a cleanup step to `Writer.write` which deletes any files in the output directory that are not in `Writer.files`.

**Comment by mscheltienne** [comment]
Hey, this PR was kind of forgotten. I really don't know much about `mffpy`, and I haven't tested the changes I made; but based on the code and on the conversion here, this last commit should address your concerns. 

Could you spend the time to finalize this one?

File: mffpy/writer.py [hunk 1]
```diff
@@ -57,6 +62,15 @@ def write(self):
...
                           xml_declaration=True, method='xml')
 
+            # remove from old files to overwrite
+            if filename in old_files:
+                idx = old_files.index(filename)
+                del old_files[idx]
+
+        # clean up in case of overwrite
+        for file in old_files:
+            remove(file)
```
Comment (ephathaway) [inline mffpy/writer.py:72]:
```suggestion
            remove(join(self.mffdir, file))
```

File: mffpy/writer.py [hunk 2]
```diff
@@ -124,7 +138,8 @@ def filename(self, fn: str):
         """check filename with .mff/.mfz extension does not exist"""
         base, ext = splitext(fn)
         assert ext in ('.mff', '.mfz', '.json')
-        assert not exists(fn), f"File '{fn}' exists already"
+        if not self.overwrite:
+            assert not exists(fn), f"File '{fn}' exists already"
         if ext == '.mfz':
             assert not exists(base + '.mff')
```
Comment (ephathaway) [inline mffpy/writer.py:144]: Let's stick this inside the previous `if` statement.

File: mffpy/tests/test_writer.py [hunk 3]
```diff
@@ -35,11 +35,12 @@ def test_writer_receives_bad_init_data():
 
 
 def test_writer_doesnt_overwrite(tmpdir):
-    """test that `mffpy.Writer` doesn't overwrite existing files"""
+    """test that `mffpy.Writer` doesn't overwrite existing files by default"""
     dirname = join(str(tmpdir), 'testdir.mff')
     makedirs(dirname, exist_ok=True)
     with pytest.raises(AssertionError, match='File.*exists already'):
         Writer(dirname)
+    Writer(dirname, overwrite=True)
```
Comment (ephathaway) [inline mffpy/tests/test_writer.py:43]: Let's make a new test for overwriting an existing MFF. It should

1. Check that the appropriate files are overwritten
2. Check that extra files are removed

This could look pretty similar to function `test_writer_writes`, but with an extra step to overwrite the first file written.

**Review by ephathaway — CHANGES_REQUESTED** [review]
Hi @mscheltienne, thanks for your contribution. With the requested changes, it should work as expected.

**Comment by mscheltienne** [comment]
@ephathaway Done

File: mffpy/writer.py [hunk 4]
```diff
@@ -57,6 +62,15 @@ def write(self):
...
                           xml_declaration=True, method='xml')
 
+            # remove from old files to overwrite
+            if filename in old_files:
+                idx = old_files.index(filename)
+                del old_files[idx]
+
+        # clean up in case of overwrite
+        for file in old_files:
+            remove(join(self.mffdir, file))
```
Comment (ephathaway) [inline mffpy/writer.py:72]: In most cases this will work. However, sometimes the MFF directory contains subdirectories. For example, in `examples/example_4.mff`, there is a `subjects` folder with a couple files. We should add some logic to remove the file if it's a file or remove the directory and all contents if it's a directory. `shutil.rmtree` should do the trick.

**Review by ephathaway — CHANGES_REQUESTED** [review]

File: mffpy/writer.py (hunk 4 above)
Comment (mscheltienne) [reply mffpy/writer.py]: done

File: mffpy/writer.py [hunk 5]
```diff
@@ -69,7 +70,10 @@ def write(self):
 
         # clean up in case of overwrite
         for file in old_files:
-            remove(join(self.mffdir, file))
+            try:
+                rmtree(join(self.mffdir, file))
+            except NotADirectoryError:
+                remove(join(self.mffdir, file))
```
Comment (jusjusjus) [inline mffpy/writer.py:76]: Why not just remove `self.mffdir`?

Comment (mscheltienne) [reply mffpy/writer.py:76]: > This sounds like a good solution. We could add a cleanup step to Writer.write which deletes any files in the output directory that are not in Writer.files.

As I am not familiar with `mffpy` and all the possible content for `self.mffdir`, I went for what was written here.

Comment (mscheltienne) [reply mffpy/writer.py:76]: You would go for `rmtree(self.mffdir)` as clean-up step if `overwrite` is set to True, instead of the file by file removal? That works for me to and is simpler.

Comment (jusjusjus) [reply mffpy/writer.py:76]: Yes, I think this might be fine.  On windows in Python 3.8 there seems to be some relevant change:

> Changed in version 3.8: On Windows, will no longer delete the contents of a directory junction before removing the junction.

If you can, please have someone test this on a windows system.

Comment (mscheltienne) [reply mffpy/writer.py:76]: > Yes, I think this might be fine.

Current version, or `rmtree(self.mffdir)` as a clean-up step?

For the change in 3.8, I'm not sure about what it means. If the junction is removed, then all the content is also removed? So it should not change anything? Do you have a CI/workflow running on windows?

Comment (mscheltienne) [reply mffpy/writer.py:76]: I can add a test where I create a directory with subdirectories in `self.mffdir` and test if it does get deleted.

Comment (ephathaway) [reply mffpy/writer.py:76]: If we're going this route, I would put the cleanup step in the `create_directory` method. If `overwrite` is `True`, `create_directory` removes all contents of the existing MFF directory, or creates a new one if it doesn't exist yet.

Comment (mscheltienne) [reply mffpy/writer.py]: done

File: mffpy/writer.py [hunk 6]
```diff
@@ -43,18 +43,16 @@ def __init__(self, filename: str, overwrite: bool = False):
     def create_directory(self):
         """Creates the directory for the recording."""
         if not self.file_created:
-            exist_ok = self.overwrite
-            makedirs(self.mffdir, exist_ok=exist_ok)
+            if self.overwrite:
+                rmtree(self.mffdir)
```
Comment (ephathaway) [inline mffpy/writer.py:47]: This results in a `FileNotFoundError` if there is no MFF directory to overwrite. If overwrite is True, the writing should be successful whether there is something to overwrite or not. So let's check that the directory exists before running `rmtree`.

File: mffpy/tests/test_writer.py [hunk 7]
```diff
@@ -75,6 +75,64 @@ def test_writer_writes(tmpdir):
...
+    W.write()
+
+    # list files
+    files = listdir(dirname)
+    assert 'info.xml' in files
+    assert 'coordinates.xml' in files
+    assert 'sensorLayout.xml' in files
+
+    # add a directory inside
+    makedirs(join(tmpdir, 'test'))
```
Comment (ephathaway) [inline mffpy/tests/test_writer.py:105]:
```suggestion
    makedirs(join(dirname, 'test'))
```

**Review by ephathaway — CHANGES_REQUESTED** [review]
So we need to account for the overwrite .mfz edge case. Currently, attempting to overwrite a .mfz file errors out because the 'mffdir' is removed (filename + .mff), but the zipped MFF is not removed (filename + .mfz). We will need to add some logic to remove both filename.mff and filename.mfz.

**Comment by mscheltienne** [comment]
Done. As there is no writer test to `.mfz`, I based my edit on your comment: 

> (filename + .mff) [...] (filename + .mfz)

**Review by ephathaway — APPROVED** [review]

**Comment by ephathaway** [comment]
@mscheltienne, thanks for your contribution! We will cut a new release with this feature.
